File: source/go/internal/core/chandramasa.go
package core

import (
	"math"

	"github.com/ishankgupta95/panchang/source/go/v5/internal/astronomy"
	"github.com/ishankgupta95/panchang/source/go/v5/internal/jsnum"
	"github.com/ishankgupta95/panchang/source/go/v5/types"
)

type BoundsAt func(refMs int64) (astronomy.NewMoonBounds, error)

func ComputeChandraMasa(
	siderealSun, siderealMoon float64,
	nameFn func(index int, isAdhika bool) string,
	system types.MasaSystem,
	refDateMs int64,
	getSiderealSun LongitudeAt,
	getBounds BoundsAt,
) (types.ChandraMasaInfo, error) {
	elongation := jsnum.Mod((siderealMoon-siderealSun)+360, 360)

	bounds, err := getBounds(refDateMs)
	if err != nil {
		return types.ChandraMasaInfo{}, err
	}
	sunAtPrevNewMoon := jsnum.Mod(jsnum.Mod(getSiderealSun(bounds.PrevMs), 360)+360, 360)
	sunAtNextNewMoon := jsnum.Mod(jsnum.Mod(getSiderealSun(bounds.NextMs), 360)+360, 360)
	solarMonthAtPrev := int(math.Floor(sunAtPrevNewMoon / 30))
	solarMonthAtNext := int(math.Floor(sunAtNextNewMoon / 30))

	isAdhika := solarMonthAtPrev == solarMonthAtNext

	amantaIndex := (solarMonthAtPrev + 1) % 12
	amantaName := nameFn(amantaIndex, isAdhika)

	isKrishnaPaksha := elongation >= 180
	purnimantaIndex := amantaIndex
	if isKrishnaPaksha && !isAdhika {
		purnimantaIndex = (amantaIndex + 1) % 12
	}
	purnimantaName := nameFn(purnimantaIndex, isAdhika)

	index, name := purnimantaIndex, purnimantaName
	if system == types.Amanta {
		index, name = amantaIndex, amantaName
	}

	return types.ChandraMasaInfo{
		Index:           index,
		Name:            name,
		IsAdhika:        isAdhika,
		System:          system,
		AmantaIndex:     amantaIndex,
		AmantaName:      amantaName,
		PurnimantaIndex: purnimantaIndex,
		PurnimantaName:  purnimantaName,
	}, nil
}

File: source/ts/src/core/chandramasa.ts
import type { ChandraMasaInfo } from '../types/elements';
import type { MasaSystem } from '../types/options';
import { boundingNewMoons, type NewMoonBounds } from '../astronomy/newMoon';

/**
 * The Amavasya initiating an Amanta month falls one rashi behind the month's name
 * (Sun in Meena → Chaitra). Mean-motion new moons overshoot the Sankranti boundary
 * near aphelion, where Adhika Jyeshtha/Ashadha occur, so the TRUE ones are required.
 */
export function computeChandraMasa(
  siderealSun: number,
  siderealMoon: number,
  nameFn: (index: number, isAdhika: boolean) => string,
  system: MasaSystem = 'purnimanta',
  refDate: Date,
  getSiderealSun: (d: Date) => number,
  getBounds: (ref: Date) => NewMoonBounds = boundingNewMoons,
): ChandraMasaInfo {
  const elongation = ((siderealMoon - siderealSun) + 360) % 360;

  const { prev, next } = getBounds(refDate);
  const sunAtPrevNewMoon = ((getSiderealSun(prev) % 360) + 360) % 360;
  const sunAtNextNewMoon = ((getSiderealSun(next) % 360) + 360) % 360;
  const solarMonthAtPrev = Math.floor(sunAtPrevNewMoon / 30);
  const solarMonthAtNext = Math.floor(sunAtNextNewMoon / 30);

  const isAdhika = solarMonthAtPrev === solarMonthAtNext;

  const amantaIndex = (solarMonthAtPrev + 1) % 12;
  const amantaName = nameFn(amantaIndex, isAdhika);

  // An Adhika month has no Sankranti, so its Purnimanta name does NOT advance.
  const isKrishnaPaksha = elongation >= 180;
  const purnimantaIndex =
    isKrishnaPaksha && !isAdhika ? (amantaIndex + 1) % 12 : amantaIndex;
  const purnimantaName = nameFn(purnimantaIndex, isAdhika);

  const index = system === 'amanta' ? amantaIndex : purnimantaIndex;
  const name = system === 'amanta' ? amantaName : purnimantaName;

  return { index, name, isAdhika, system, amantaIndex, amantaName, purnimantaIndex, purnimantaName };
}
